### PR TITLE
Nullspace example

### PR DESCRIPTION
**Pull request by snschune**
@fdkong @YaqiWang @jwpeterson are interested. 

I provide an example to show a possible issue with the nullspace capability in Moose. 

If you execute in serial or with multiple threads, you get the correct answer (answer == values of u at the nodes that get through a VectorPP). If you execute 
on two processors, the answer is incorrect. 

I verified the answer against the analytical answer so I know it's correct when running with 
in serial or just multi-threading. I have a python script that computes and checks the right
answer in case anyone wants to take a close look.

### CODE REVIEW

**Comment by YaqiWang** [comment]
Thanks @snschune This is a nice test setup.

I think moose should be enhanced to make the null space setup easier and more useful. For example, from the initial solution, or from intial aux solution, or from a function, etc.

**Review by fdkong — COMMENTED** [review]
> Thanks @snschune This is a nice test setup.
> 
> I think moose should be enhanced to make the null space setup easier and more useful. For example, from the initial solution, or from intial aux solution, or from a function, etc.

Null spaces could be any thing, and it totally determined by specific applications.  There is no a general way to do this.

File: test/src/executioners/SteadyWithNull.C [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    Steady(parameters)
+{
+}
+
+void
+SteadyWithNull::init()
+{
+  Steady::init();
+  NumericVector<Number> * to_vector = &_problem.getNonlinearSystem().getVector("NullSpace_0");
+  const NumericVector<Number> * from_vector = &_problem.getAuxiliarySystem().solution();
```
Comment (YaqiWang) [inline test/src/executioners/SteadyWithNull.C]: so we should use `problem.getAuxiliarySystem().currentSolution()?`

**Comment by YaqiWang** [comment]
Understand that. The problem now is that even the user knows what the null space is, he cannot set it in the input file. What I am proposing is to enable this.

**Comment by fdkong** [comment]
> Understand that. The problem now is that even the user knows what the null space is, he cannot set it in the input file. What I am proposing is to enable this.

@snschune, @YaqiWang, Could you possibly also create an issue? I will come back to this after I finished something.

**Comment by YaqiWang** [comment]
See #7780. That issue can be a little changed to be more precise.

File: test/src/executioners/SteadyWithNull.C (hunk 1 above)
Comment (snschune) [inline test/src/executioners/SteadyWithNull.C]: I think, this is exactly the same in the MultiAppNumericVector transfer we have in Rattlesnake. 
So if the above code causes the issue, the same I believe would be in our transfer.

File: test/src/executioners/SteadyWithNull.C [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  *to_vector = *from_vector;
```
Comment (snschune) [inline test/src/executioners/SteadyWithNull.C]: Do you mean value by value? Isn't this just taking the content of one vector and assigns it to the other?

**Comment by fdkong** [comment]
Got the same solution with using one, two, and four processors, @YaqiWang. Possibly it is related to PETSc version? I am using PETSc-3.7.3.

**Comment by YaqiWang** [comment]
Can hardly believe it because PETSc is pretty stable. @fdkong can you verify if there recently is a bug fix in the null space stuff in PETSc? You might want to try it on on Jason's build box by turning this to a moose test, where 3.6.4 is installed i think.

**Comment by fdkong** [comment]
Using precondiitoners such as lu, and hypre resolves this issue. I am going to close this PR. How do you guys think, @YaqiWang, @snschune ?

**Comment by snschune** [comment]
This should become a regular test. I'll add a test and gold files.

**Comment by snschune** [comment]
@fdkong The second commit has an example where the rhs has a contamination that is outside the range of the matrix `A` if we solve `Ax = b`. 

Note that in this case (Laplacian operator) the matrix `A` is symmetric and that's why I know that the complement of the range of `A` is spanned by the fundamental mode which I then add. 
In the `contaminated_second_harmonic` function you can modify `a`. Set it to zero means the rhs is within the range of `A` and you converge, but if you set it to anything larger than 0 you'll see you only converge to something of the order of `a` in nonlinear residual norm. 

imho that indicates that the nullspace of the transpose is not correctly removed by PETSc.

**Comment by fdkong** [comment]
> @fdkong The second commit has an example where the rhs has a contamination that is outside the range of the matrix A if we solve Ax = b.
> 
> Note that in this case (Laplacian operator) the matrix A is symmetric and that's why I know that the complement of the range of A is spanned by the fundamental mode which I then add.
> In the contaminated_second_harmonic function you can modify a. Set it to zero means the rhs is within the range of A and you converge, but if you set it to anything larger than 0 you'll see you only converge to something of the order of a in nonlinear residual norm.
> 
> imho that indicates that the nullspace of the transpose is not correctly removed by PETSc.

I think what users should do is (1) to provide kernel(A) via orthogonal basis vectors (2) to make sure b in range(A).

**Comment by fdkong** [comment]
Clicked on the wrong button. Sorry.

**Comment by snschune** [comment]
I'll add these tests with another PR that also adds the transpose interface.

**Pull request closed without merging**